core/src/main/scala/kafka/controller/ControllerContext.scala

```diff
@@ -431,3 +425,22 @@ class ControllerContext {
     targetState.validPreviousStates.contains(partitionStates(partition))
 
 }
+
+/**
+  * The ControllerContextSnapshot is an immutable snapshot of the ControllorContext.
+  * The motivation for this class is that we don't need to calculate certain fields
+  * repeatedly, including liveBrokerIds and liveOrShuttingDownBrokerIds.
+  */
+case class ControllerContextSnapshot(controllerContext: ControllerContext) {
+  val liveBrokerIds = controllerContext.liveBrokerIds
+  val liveOrShuttingDownBrokerIds = controllerContext.liveOrShuttingDownBrokerIds
+  val replicasOnOfflineDirs = controllerContext.replicasOnOfflineDirs
```

Comment:
So all these fields are computed?  It's not clear because Scala is hiding method call vs field dereference.

Reply:
The first 2 are method calls, and the last one `replicasOnOfflineDirs` is a field.
I'm captured all 3 here since they are needed for the isReplicaOnline call.